Fix pyenv setup for fish

relaxolotl opened the pull request:
I bumped into a few issues while setting up the environment while using fish as my shell, so this fixes those up and performs a bit of additional cleanup.

Setup script updates: https://github.com/getsentry/sentry/pull/26511

relaxolotl commented on src/docs/environment/index.mdx:
`````diff
@@ -213,16 +257,25 @@ And after doing that, reload your shell:
 
-Two examples of variables that you can place there are `SENTRY_SILENCE_DIRENV_NOTICE` to make `direnv` output less noisy
-and `SENTRY_DEVENV_NO_REPORT=1` to prevent the development environment to report errors to Sentry.
+Some useful variables to set in `.env`:
+
+- `SENTRY_SILENCE_DIRENV_NOTICE`: Make `direnv` output less noisy.
+- `SENTRY_DEVENV_NO_REPORT`: Do not report development environment errors to Sentry.io.
+- `SENTRY_DIRENV_DEBUG`: Print debug messages to stdout.
+
+A good starter `.env`:
+```shell {filename: .env}
+SENTRY_SILENCE_DIRENV_NOTICE=1
+SENTRY_DEVENV_NO_REPORT=1
+````
`````
Not all of the values supported in .env are documented anywhere, so this should serve as a
starting point in documenting these settings. An example .env has also been added to
improve ease of use. Just create the file, copy-paste the example, and you're set!

armenzg replied:
Maybe we can create a `.env.example` in Sentry and we slowly add all variables with their *default* values.
Or we even `cp .env.example .env` as part of the bootstrap script if the file does not exist.

What do you think?

relaxolotl replied:
I like that idea! To build onto that, I'm thinking we could also omit the explanations for the variables by including descriptions for them in the `.env.example` itself, assuming it's possible to include comments there. If not, then I can keep this list.

Our suggested defaults appear to include all of the supported variables right now, so once a non-default variable is introduced we could begin exploring how we would like to document those in this repository. Thoughts?

armenzg replied:
For optional aspects of the dev env, I would like to keep it directly via the CLI in Sentry. There's no easy way of keeping the develop docs in sync with what is happening in Sentry.

relaxolotl commented on src/docs/environment/index.mdx:
````diff
@@ -89,23 +89,55 @@ computer is actually compiling Python!
 again and cd into sentry. To reload it, run:
 
-```shell
+```shell {tabTitle: Bash}
+exec "$SHELL"
+```
+
+```shell {tabTitle: Zsh}
 exec "$SHELL"
 ```
 
+```shell {tabTitle: Fish}
+exec $SHELL
+```
````
This change was made in a few other additional places in the PR.

It's less code overall to just mention fish's differences in text, but it's also additional
required reading for the end user. Stuffing these into code blocks provides a more streamlined
experience since their selected languages sync across an entire page, meaning that once a user
decides to view the fish instructions in one code block, the others will switch to displaying
fish as well.

relaxolotl commented on src/docs/environment/index.mdx:
`````diff
@@ -89,23 +89,55 @@ computer is actually compiling Python!
+fi
+````
 
-Note: If you use Fish, add `source (pyenv init - | psub)` to `~/.config/fish/config.fish`.
`````
See: pyenv's instructions in their [README](https://github.com/pyenv/pyenv#basic-github-checkout)

relaxolotl commented on src/docs/environment/index.mdx:
````diff
@@ -89,23 +89,55 @@ computer is actually compiling Python!
 make setup-pyenv
 ```
 
+fish users will need to manually set some of the environment variables. This only needs to be done
+once.
+
+```shell {tabTitle: Fish}
+set -Ux PYENV_ROOT $HOME/.pyenv
+# fish >=3.2.0
+fish_add_path $PYENV_ROOT/bin
+# fish <3.2.0
+set -U fish_user_paths $PYENV_ROOT/bin $fish_user_paths
+```
+
````
It's possible to insert lines equivalent to the interactively run commands into the 
startup script instead, but it's difficult to get this right. A common fish mistake is to append the 
pyenv directory without any regard for whether it's already present in `PATH` or not, causing 
it to explode in length based on how often the user spins up new shells. 

In comparison, the fish commands proposed in this PR only need to be run once, will properly append to the `PATH`, and therefore should not cause too much developer pain.

relaxolotl commented on src/docs/environment/index.mdx:
```diff
@@ -213,16 +257,25 @@ And after doing that, reload your shell:
-and `SENTRY_DEVENV_NO_REPORT=1` to prevent the development environment to report errors to Sentry.
+Some useful variables to set in `.env`:
+
+- `SENTRY_SILENCE_DIRENV_NOTICE`: Make `direnv` output less noisy.
```
I actually can't seem to find any evidence of this variable's existence in sentry nor getsentry. Does this still exist?

armenzg replied:
You're right. It's fallen out of date.

relaxolotl replied:
I'll remove this from the diff. 👍

armenzg commented on src/docs/environment/index.mdx:
````diff
@@ -89,23 +89,55 @@ computer is actually compiling Python!
-```shell
+
+```shell {filename: ~/.bash_profile} {tabTitle: Bash}
+eval "$(pyenv init --path)"
````
I read the bash instructions and I decided to leave it to the user to follow the instructions (besides most engineers don't use bash).
```suggestion
Visit https://github.com/pyenv/pyenv#installation for instructions on how to set up Bash.
```

armenzg commented on src/docs/environment/index.mdx:
`````diff
@@ -89,23 +89,55 @@ computer is actually compiling Python!
+````
+
+```shell {filename: ~/.zshrc} {tabTitle: Zsh}
+eval "$(pyenv init --path)"
`````
```suggestion
# Since we have installed via Brew we only need to do this
echo 'eval "$(pyenv init --path)"' >> ~/.zprofile
```

relaxolotl replied:
I've opted to keep the second line as is (`eval "$(pyenv init --path)"`) since this is meant to represent what gets inserted into the user's `.zprofile`. It's subtle, but the filename portion of the codeblock should also provide context on where exactly this gets added.

armenzg commented on src/docs/environment/index.mdx:
````diff
@@ -89,23 +89,55 @@ computer is actually compiling Python!
+
+```shell {filename: ~/.zshrc} {tabTitle: Zsh}
+eval "$(pyenv init --path)"
+if command -v pyenv 1>/dev/null 2>&1; then
````
```suggestion
```

armenzg commented on src/docs/environment/index.mdx:
````diff
@@ -89,23 +89,55 @@ computer is actually compiling Python!
+```shell {filename: ~/.zshrc} {tabTitle: Zsh}
+eval "$(pyenv init --path)"
+if command -v pyenv 1>/dev/null 2>&1; then
+  eval "$(pyenv init -)"
````
```suggestion
```

armenzg commented on src/docs/environment/index.mdx:
```diff
@@ -89,23 +89,55 @@ computer is actually compiling Python!
+eval "$(pyenv init --path)"
+if command -v pyenv 1>/dev/null 2>&1; then
+  eval "$(pyenv init -)"
+fi
```
```suggestion
```

armenzg commented on src/docs/environment/index.mdx:
````diff
@@ -135,17 +175,24 @@ curl https://get.volta.sh | bash
 
+```shell {tabTitle: Fish}
+exec $SHELL
+```
````
Are you sure this is better since the three are exactly the same?

I can see *some* value if they ever were to differ.

armenzg commented on src/docs/environment/index.mdx:
```diff
@@ -213,16 +257,25 @@ And after doing that, reload your shell:
 
-Two examples of variables that you can place there are `SENTRY_SILENCE_DIRENV_NOTICE` to make `direnv` output less noisy
-and `SENTRY_DEVENV_NO_REPORT=1` to prevent the development environment to report errors to Sentry.
+Some useful variables to set in `.env`:
```
We now have a make target that gives us the latest doc. Would you mind deleting the other lines and just keeping this one I suggest?

```suggestion
If you execute `make direnv-help` you will see the latest env variables to configure direnv to your taste.
```

relaxolotl replied:
I've taken slight liberties in updating this; Since we're [playing with the idea of having a template `.env` file](https://github.com/getsentry/sentry/pull/26530) I've made note of that in this PR. `make direnv-help` does get brought up, though.

armenzg reviewed (changes requested):
Thank you for this PR and getting fish into a better state!

Few small changes, otherwise, it looks good!

relaxolotl commented on src/docs/environment/index.mdx:
```diff
@@ -213,16 +234,17 @@ And after doing that, reload your shell:
 
-Two examples of variables that you can place there are `SENTRY_SILENCE_DIRENV_NOTICE` to make `direnv` output less noisy
-and `SENTRY_DEVENV_NO_REPORT=1` to prevent the development environment to report errors to Sentry.
+One of `bootstrap`'s steps [creates a starter `.env`](https://github.com/getsentry/sentry/blob/master/.env.example) which should 
```
This link points to a nonexistent destination until https://github.com/getsentry/sentry/pull/26530 is merged in.

relaxolotl commented:
Thank you again @armenzg for also reviewing this! I believe I've addressed all comments, and any slight changes are mentioned in replies to your comments. This should be good for a re-review.

armenzg commented on src/docs/environment/index.mdx:
`````diff
@@ -89,23 +89,40 @@ computer is actually compiling Python!
 ```shell
 exec "$SHELL"
-```
+````
`````
Do not make this change (3 single quotes is the right way).

relaxolotl replied:
Thanks for catching this, looks like I copy-pasted a bad change to a couple of places 🙃

armenzg commented on src/docs/environment/index.mdx:
`````diff
@@ -89,23 +89,40 @@ computer is actually compiling Python!
-```
+```shell {tabTitle: Bash}
+Given that the bash instructions vary greatly based on the user's configuration, it is recommended to visit https://github.com/pyenv/pyenv#installation for instructions on how to set up Bash.
+````
`````
Fix 4 single quotes.

armenzg commented on src/docs/environment/index.mdx:
`````diff
@@ -89,23 +89,40 @@ computer is actually compiling Python!
+```shell {filename: ~/.zprofile} {tabTitle: Zsh}
+# It is assumed that pyenv is installed via Brew, so this is all we need to do.
+eval "$(pyenv init --path)"
+````
`````
4 single quotes.

armenzg commented on src/docs/environment/index.mdx:
`````diff
@@ -137,15 +162,14 @@ reload your shell:
 ```shell
 exec "$SHELL"
-```
+````
`````
4 single quotes.

armenzg reviewed (changes requested):
Some more minor changes.
